src/proof/arith_proof.cpp

```diff
@@ -833,26 +840,317 @@ void LFSCArithProof::printOwnedSort(Type type, std::ostream& os) {
   }
 }
 
-void LFSCArithProof::printTheoryLemmaProof(std::vector<Expr>& lemma, std::ostream& os, std::ostream& paren, const ProofLetMap& map) {
-  os << " ;; Arith Theory Lemma \n;;";
-  for (unsigned i = 0; i < lemma.size(); ++i) {
-    os << lemma[i] <<" ";
+void LFSCArithProof::printRational(std::ostream& o, const Rational& r)
+{
+  if (r.sgn() < 0)
+  {
+    o << "(~ " << r.getNumerator().abs() << "/" << r.getDenominator().abs()
+      << ")";
+  }
+  else
+  {
+    o << r.getNumerator() << "/" << r.getDenominator();
+  }
+}
+
+void LFSCArithProof::printLinearPolynomialNormalizer(std::ostream& o, const Node& n)
+{
+  switch (n.getKind())
+  {
+    case kind::PLUS:
+      {
+        // Since our axioms are binary, but n may be n-ary, we rig up
+        // a right-associative tree.
+        for (size_t i = 0; i < n.getNumChildren(); ++i)
```

Comment:
I'm happy to make this change, and have done so.

However, for my own learning: why is this necessary? I would have thought that this is an easy optimization for the compiler to spot (e.g. in [this](https://godbolt.org/z/S7zeKY) situation the compiler seems to do it).

Reply:
It can not always be optimized away, e.g., [here](https://godbolt.org/z/IPiWaT).

As a rule of thumb: If it is not absolutely guaranteed, make sure it is not an issue.

Since it is not guaranteed here in general, we do not rely on the compiler but make sure that it can't happen. What is maybe also worth mentioning, it is in general always better to temporarily store heap values you need very often (in functions that are called very often) in a stack variable rather than having to dereference and access the heap every time. Things like this can make quite a performance impact in SMT solvers since we deal with internal representations of formulas that can be huge (lots of nodes, thus lots of function calls if you, e.g., traverse the structure, even worse if you do it again and again, which is very common).

Reply:
Point taken.

As a note though, the optimization didn't happen in your example only because the optimizations were not enabled.